url696d9a93c9_liaosiwei_bayesnetwork_tgz-pJ8-tutorial_TutorialJ8/src/org/eclipse/recommenders/jayes/inference/LikelihoodWeightedSampling.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (c) 2011 Michael Kutschke.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Michael Kutschke - initial API and implementation.
+ */
+package org.eclipse.recommenders.jayes.inference;
+
+import java.util.Map;
+
+import org.eclipse.recommenders.jayes.BayesNet;
+import org.eclipse.recommenders.jayes.BayesNode;
+import org.eclipse.recommenders.jayes.sampling.BasicSampler;
+import org.eclipse.recommenders.jayes.util.MathUtils;
+
+public class LikelihoodWeightedSampling extends AbstractInferer {
+
+    private int sampleCount = 200;
+    private BasicSampler sampler = new BasicSampler();
+
+    @Override
+    public void setNetwork(BayesNet bn) {
+        super.setNetwork(bn);
+        sampler.setNetwork(bn);
+
+    }
+
+    @Override
+    protected void updateBeliefs() {
+        sampler.setEvidence(evidence);
+        for (int i = 0; i < sampleCount; i++) {
+            Map<BayesNode, String> sample = sampler.sample();
+            double weight = computeEvidenceProbability(sample);
+
+            for (BayesNode e : sample.keySet()) {
+                beliefs[e.getId()][e.getOutcomeIndex(sample.get(e))] += weight;
+            }
+        }
+
+        normalizeBeliefs();
+
+    }
+
+    private void normalizeBeliefs() {
+        for (int i = 0; i < beliefs.length; i++)
+            beliefs[i] = MathUtils.normalize(beliefs[i]);
+    }
+
+    private double computeEvidenceProbability(Map<BayesNode, String> sample) {
+        double factor = 1.0;
+        for (BayesNode n : evidence.keySet()) {
+            factor *= n.marginalize(sample)[n.getOutcomeIndex(evidence.get(n))];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.